mlir/lib/Dialect/Rise/EDSC/HighLevel.cpp

```diff
@@ -31,22 +31,73 @@ using namespace mlir::edsc::type;
 using namespace mlir::edsc::abstraction;
 using namespace mlir::edsc::intrinsics;
 
-// void mlir::edsc::highlevel::makeRiseProgram(ArrayRef<Value> inputs, Value
-// output, function_ref<Value(BlockArgument)> bodyBuilder) {
-//  return;
-//}
-//
+Type inferTypeForInOp(Value input) {
+  if (MemRefType inMemrefType = input.getType().dyn_cast<MemRefType>()) {
+    // infer input type from memref
+    DataType inputType = scalarF32Type();
+    for (auto size = inMemrefType.getShape().rbegin();
+         size != inMemrefType.getShape().rend(); size++) {
+      inputType = arrayType(*size, inputType);
+    }
+    return inputType;
+  }
+  emitError(input.getLoc())
+      << "Could not infer type for rise.in operation with provided input!";
+  return nullptr;
+}
+
+void mlir::edsc::highlevel::makeRiseProgram(
```

Comment:
I spent some time trying to template these functions but ran into some issues. 
The variadic argument has to be the last one in the signature, which would look like this: `makeRiseProgram(Value output, function_ref<Value(Value, Args...)> bodyBuilder, Value input, Args... inputs);` 
This is very unhandy with the inputs in the back.
Defining the base case and the case with multiple inputs works alright, but the function with multiple inputs can never be called, as the pattern matching with the lambda does not work. 
I really want to avoid the type of the bodybuiler being `function_ref<Value(ArrayRef<Value>)> bodyBuilder` as that is way less comfortable when writing the rise program.

Reply:
You can do something like this:

```
#include <functional>
#include <tuple>

struct Value {

};

template <typename T, typename... Targs>
struct builder {
    Value output;
    std::tuple<T, Targs...> inputs;
    builder(Value output, std::tuple<T, Targs...> inputs)
        : output{output}, inputs{inputs}
    {}

    void operator()(std::function<Value(T, Targs...)> fun) {

    }
};

template <typename T, typename... Targs>
builder<T, Targs...> makeRiseProgram(Value output, T input, Targs... inputs) {
    return builder<T, Targs...>{ output, std::make_tuple(input, inputs...) };
}


int main() {
    auto output = Value{};
    auto input = Value{};

    makeRiseProgram(output, input)([](auto v1) { return v1; } );
    makeRiseProgram(output, input, input)([](auto v1, auto v2) { return v1; } );
}
```